### PR TITLE
Fix importing Spack packages as Python modules

### LINKED ISSUE #29203: Swapping repositories creates new module objects in `spack.pkg.*`
### Steps to reproduce

```console
$ spack python
Spack version 0.17.1
Python 3.8.10, Linux x86_64
>>> import spack.repo
>>> zlib1 = spack.repo.path.get_pkg_class('zlib')
>>> zlib2 = spack.repo.path.get_pkg_class('zlib')
>>> id(zlib1)
41925296
>>> id(zlib2)
41925296
>>> import spack.paths
>>> with spack.repo.use_repositories(spack.paths.packages_path):
...     zlib3 = spack.repo.path.get_pkg_class('zlib')
...     id(zlib3)
... 
42258432
```
The class id is different because the module object has been recreated from scratch, even though one was already available in `sys.modules`.

### Error message

No error message, just the unexpected behavior above which might lead to intricate failures when two different module objects (which are supposed to be equal) get used in different parts of the code.

### Information on your system

* **Spack:** 0.17.1-1357-d62b8f0bf3
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by alalazo**
fixes #29203

This PR fixes a subtle bug we have when importing Spack packages as Python modules that can lead to multiple module objects being created for the same package. It also fixes all the places in unit-tests where relying on the old bug was crucial to have a new "clean" state of the package class.

The gist of the fix is:

https://github.com/alalazo/spack/blob/815557ffb1f598130453645e62b74b6b6a6bd1ca/lib/spack/spack/repo.py#L1104-L1108

where we go through the Python importing protocol to retrieve the package classes instead of reloading them from sources again and again.

### CODE REVIEW

**Comment by alalazo** [comment]
~Tests seems to run much faster with this fix (they don't have to reload from packages sources multiple times).~ Turns out it's a bug, fixed by #29223

**Comment by alalazo** [comment]
@scheibelp @becker33 Pinging you both since we have been discussing this yesterday

File: lib/spack/spack/test/install.py [hunk 1]
```diff
@@ -160,9 +162,11 @@ def test_dont_add_patches_to_installed_package(install_mockery, mock_fetch):
     dependent = Spec('dependent-install ^/' + dependency_hash)
     dependent.concretize()
 
-    dependency.package.patches['dependency-install'] = [
+    monkeypatch.setitem(dependency.package.patches, 'dependency-install', [
         spack.patch.UrlPatch(
-            dependent.package, 'file://fake.patch', sha256='unused-hash')]
+            dependent.package, 'file://fake.patch', sha256='unused-hash'
+        )
+    ])
```
Comment (becker33) [inline lib/spack/spack/test/install.py:169]: Shouldn't we not need to do this because it's affecting the package object, not the package class? Same applies to the other fixes to use monkeypatch on `Package` objects.

Comment (alalazo) [reply lib/spack/spack/test/install.py:169]: No, it's affecting the class, since it's setting an item in a class-level dictionary which was already present.

**Comment by alalazo** [comment]
The failure in tests happens only with the original concretizer when executing these two tests:
```console
$ SPACK_TEST_SOLVER=original bin/spack unit-test lib/spack/spack/test/concretize.py::TestConcretize::test_patching_dependencies lib/spack/spack/test/patch.py::test_nested_directives
```
since the first one, for some reason I still have to pin, the first test modifies the class-level dictionary for the second class.

**Comment by alalazo** [comment]
The issue with the original concretizer is that we modify the class level package dictionary here:

https://github.com/spack/spack/blob/b20df12d09b190798837c6cb9e70fb8e16565d75/lib/spack/spack/dependency.py#L124-L129

This is not an issue with `clingo` since we don't have to keep track of patches during a "normalize" phase (the conditions are expressed in a logic program that is solved as a whole and generate an answer set). With the original concretizer we were implicitly relying on the current bug we had to "reset" the package class.

**Comment by alalazo** [comment]
@becker33 This is ready for review. The lines of code changed in core Spack are just a few, but they are fundamental to loading packages. Let me know if you see any lack of testing with respect to that.

**Comment by scheibelp** [comment]
Regarding https://github.com/spack/spack/pull/29221#issuecomment-1055221077

I think the issue is: the `Dependency` objects are managed by the `Package` class, and so multiple concretizations can repeatedly append more patches (so tests looking for exact list-based matching would fail as duplicates were added).

I would think there would be another issue though: a later concretization of a different Spec with the old concretizer might now exclude patches (e.g. because conditions are no longer met) - is this a problem (and none of our tests catch it) or am I missing something?

**Comment by alalazo** [comment]
> I would think there would be another issue though: a later concretization of a different Spec with the old concretizer might now exclude patches (e.g. because conditions are no longer met) - is this a problem (and none of our tests catch it) or am I missing something?

I'm not sure I understand your point, but I think this is not an issue. My understanding is the following. Patches are recorded in 2 places:
1. Package class `patches` attribute (the most part, for patches affecting the same package)
2. Dependent package class `dependencies` attribute values (in the case of packages that impose a patch on a dependency)

It seems to me that the bug solved here triggered another latent bug in case number 2.:

https://github.com/spack/spack/blob/b20df12d09b190798837c6cb9e70fb8e16565d75/lib/spack/spack/dependency.py#L124-L129

where we were extending a list without checking if the elements were duplicate or not. This has been solved here by defining what does it mean for a patch to be equal to another (they are equal if have the same sha256) and avoiding extending the list with entries that are already present.

**Comment by scheibelp** [comment]
To rephrase my concern: say you have two packages `X->Y` (X depends on Y), and lets say that `X+foo` applies a patch to `Y`, but `X~foo` does not.

If we try to concretize both, we may end up recording a patch for `X~foo->Y` when we shouldn't.

I should say that it looks like we don't allow for conditional patches on dependencies, so if `X->Y` with some patch on Y, then we will always apply the patch to Y. So my original concern is not an issue. But does that clarify what my original concern was?

**Comment by alalazo** [comment]
> I should say that it looks like we don't allow for conditional patches on dependencies, so if X->Y with some patch on Y, then we will always apply the patch to Y. So my original concern is not an issue. But does that clarify what my original concern was?

Thanks, that does clarify the point. My understanding is that we take care of that case already. What is stored in packages is always a dictionary that maps a condition being met to a patch to be applied. So we can do:
```
depends_on('libelf')
depends_on('libelf', patches='foo.patch', when='+foo')
```
and have the patch only when the spec is `+foo`. For instance using the following diff:
```diff
diff --git a/var/spack/repos/builtin.mock/packages/patch-several-dependencies/package.py b/var/spack/repos/builtin.mock/packages/patch-several-dependencies/package.py
index f8372c4bed..d3c24611a0 100644
--- a/var/spack/repos/builtin.mock/packages/patch-several-dependencies/package.py
+++ b/var/spack/repos/builtin.mock/packages/patch-several-dependencies/package.py
@@ -21,8 +21,9 @@ class PatchSeveralDependencies(Package):
     # demonstrate all the different ways to patch things
 
     # single patch file in repo
-    depends_on('libelf', patches='foo.patch')
-    depends_on('libelf@0.8.10', patches='foo.patch', when='+foo')
+    depends_on('libelf')
+    depends_on('libelf', patches='foo.patch', when='+foo')
+    #depends_on('libelf@0.8.10', patches='foo.patch', when='+foo')
 
     # using a list of patches in one depends_on
     depends_on('libdwarf', patches=[

```
we can:
```
$ spack -m python
Spack version 0.17.1
Python 3.8.10, Linux x86_64
>>> import spack.spec
>>> s1 = spack.spec.Spec('patch-several-dependencies+foo')
>>> s2 = spack.spec.Spec('patch-several-dependencies~foo')
>>> s1._old_concretize()
==> Warning: the original concretizer is currently being used.
	Upgrade to "clingo" at your earliest convenience. The original concretizer will be removed from Spack starting at v0.18.0
>>> s2._old_concretize()
==> Warning: the original concretizer is currently being used.
	Upgrade to "clingo" at your earliest convenience. The original concretizer will be removed from Spack starting at v0.18.0
>>> s1
patch-several-dependencies@2.0%gcc@9.3.0+foo arch=linux-ubuntu20.04-icelake ^fake@1.0%gcc@9.3.0 patches=1234abc,abcd123 arch=linux-ubuntu20.04-icelake ^libelf@0.8.13%gcc@9.3.0 patches=b5bb9d8 arch=linux-ubuntu20.04-icelake
>>> s2
patch-several-dependencies@2.0%gcc@9.3.0~foo arch=linux-ubuntu20.04-icelake ^fake@1.0%gcc@9.3.0 patches=1234abc,abcd123 arch=linux-ubuntu20.04-icelake ^libelf@0.8.13%gcc@9.3.0 arch=linux-ubuntu20.04-icelake
```
where we see the patch is correctly applied only to `s1`.

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
Thanks for the explanation - I missed that we were storing separate `Dependency` objects for different conditions. I have approved and would be fine with merging, but I notice you specifically pinged @becker33 so I'll wait on that (you can let me know if you want me to merge it though, or merge it yourself).

**Comment by alalazo** [comment]
> I notice you specifically pinged @becker33 so I'll wait on that (you can let me know if you want me to merge it though, or merge it yourself)

I pinged @becker33 since he was the first to review, but I think one approval is fine for this PR. I'll go on merging and monitor if we see anything odd being reported.
